mcp/src/main/java/io/modelcontextprotocol/client/auth/OAuthClientProvider.java

```diff
@@ -0,0 +1,594 @@
+		String authBaseUrl = getAuthorizationBaseUrl(serverUrl);
+		String url = authBaseUrl + "/.well-known/oauth-authorization-server";
+
+		HttpRequest request = HttpRequest.newBuilder()
+			.uri(URI.create(url))
+			.header("MCP-Protocol-Version", "0.1")
+			.GET()
+			.build();
+
+		return httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString()).thenApply(response -> {
+			if (response.statusCode() == 404) {
+				return null;
+			}
+			if (response.statusCode() != 200) {
+				throw new RuntimeException("Failed to discover OAuth metadata: " + response.statusCode());
+			}
+			try {
+				return objectMapper.readValue(response.body(), OAuthMetadata.class);
+			}
+			catch (IOException e) {
+				throw new RuntimeException("Failed to parse OAuth metadata", e);
+			}
+		}).exceptionally(ex -> {
+			// Try again without MCP header
+			HttpRequest retryRequest = HttpRequest.newBuilder().uri(URI.create(url)).GET().build();
+
+			try {
+				HttpResponse<String> response = httpClient.send(retryRequest, HttpResponse.BodyHandlers.ofString());
+				if (response.statusCode() == 404) {
+					return null;
+				}
+				if (response.statusCode() != 200) {
+					return null;
+				}
+				return objectMapper.readValue(response.body(), OAuthMetadata.class);
+			}
+			catch (Exception e) {
+				return null;
+			}
+		});
+	}
+
+	/**
+	 * Get or register client with server.
+	 * @return A CompletableFuture that resolves to the client information.
+	 */
+	private CompletableFuture<OAuthClientInformation> getOrRegisterClient() {
+		if (clientInfo != null) {
+			return CompletableFuture.completedFuture(clientInfo);
+		}
+
+		return registerOAuthClient(serverUrl, clientMetadata, metadata).thenCompose(registeredClient -> {
+			this.clientInfo = registeredClient;
+			return storage.setClientInfo(registeredClient).thenApply(v -> registeredClient);
+		});
+	}
+
+	/**
+	 * Register OAuth client with server.
+	 * @param serverUrl The server URL.
+	 * @param clientMetadata The client metadata.
+	 * @param metadata The OAuth metadata.
+	 * @return A CompletableFuture that resolves to the registered client information.
+	 */
+	private CompletableFuture<OAuthClientInformation> registerOAuthClient(String serverUrl,
+			OAuthClientMetadata clientMetadata, OAuthMetadata metadata) {
+
+		String registrationUrl;
+		if (metadata != null && metadata.getRegistrationEndpoint() != null) {
+			registrationUrl = metadata.getRegistrationEndpoint().toString();
+		}
+		else {
+			// Use fallback registration endpoint
+			String authBaseUrl = getAuthorizationBaseUrl(serverUrl);
+			registrationUrl = authBaseUrl + "/register";
+		}
+
+		// Handle default scope
+		if (clientMetadata.getScope() == null && metadata != null && metadata.getScopesSupported() != null
+				&& !metadata.getScopesSupported().isEmpty()) {
+			clientMetadata.setScope(String.join(" ", metadata.getScopesSupported()));
+		}
+
+		try {
+			String requestBody = objectMapper.writeValueAsString(clientMetadata);
```

Comment:
Testing with mcp.wix.com I had to add `.setSerializationInclusion(JsonInclude.Include.NON_NULL)` and `.setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)`. The server did not accept optional fields with null values, and `redirectUris` had to be `redirect_uris`. Ideally we just `@JsonProperty` annotations to the `OAuthClientMetadata` object according to the spec.